test-tests/shared/src/test/scala/zio/test/CheckSpec.scala

```diff
@@ -0,0 +1,93 @@
+package zio.test
+
+import zio.test.Assertion._
+import zio.test.TestAspect.failure
+import zio.test.TestUtils.execute
+import zio.test.{ check => Check, checkM => CheckM }
+import zio.{ random, Chunk, Ref, ZIO }
+
+object CheckSpec
+    extends ZIOBaseSpec(
+      suite("CheckSpec")(
+        testM("checkM is polymorphic in error type") {
+          CheckM(Gen.int(1, 100)) { n =>
+            for {
+              _ <- ZIO.effect(())
+              r <- random.nextInt(n)
+            } yield assert(r, isLessThan(n))
+          }
+        },
+        testM("effectual properties can be tested") {
+          CheckM(Gen.int(1, 100)) { n =>
+            for {
+              r <- random.nextInt(n)
+            } yield assert(r, isLessThan(n))
+          }
+        },
+        testM("error in checkM is test failure") {
+          CheckM(Gen.int(1, 100)) { n =>
+            for {
+              _ <- ZIO.fail("fail")
+              r <- random.nextInt(n)
+            } yield assert(r, isLessThan(n))
+          }
+        } @@ failure,
+        testM("overloaded check methods work") {
+          Check(Gen.anyInt, Gen.anyInt, Gen.anyInt) { (x, y, z) =>
+            assert((x + y) + z, equalTo(x + (y + z)))
+          }
+        },
+        testM("max shrinks is respected") {
+          val gen = Gen.listOfN(10)(Gen.int(-10, 10))
+          for {
+            ref <- Ref.make(0)
+            _ <- execute {
```

Comment:
You can simplify this by removing the `execute` and `testM` calls. Since `checkM` itself returns an effect with a test result you can just include that in your for comprehension directly.

Reply:
got it, will try, thanks